src/compiler/parser.ts

```diff
@@ -2615,6 +2615,7 @@ namespace ts {
         }
 
         function parseUnionOrIntersectionType(kind: SyntaxKind, parseConstituentType: () => TypeNode, operator: SyntaxKind): TypeNode {
+            parseOptional(operator);

```

Comment:
You need to also add `SyntaxKind.BarToken` and `SyntaxKind.AmpersandToken` to the `isStartOfType` function.